### PR TITLE
metrics: smgr query time: add a pre-aggregated histogram

### PR DESCRIPTION
**Pull request by problame**
When doing global queries in VictoriaMetrics, the per-timeline
histograms make us run into cardinality limits.

We don't want to give them up just yet because we don't
have an alternative for drilling down on timeline-specific
performance issues.

So, add a pre-aggregated histogram and add observations to it
whenever we add observations to the per-timeline histogram.

While we're at it, switch to using a strummed enum for the operation
type names.

### CODE REVIEW

File: pageserver/src/metrics.rs [hunk 1]
```diff
@@ -570,23 +570,157 @@ pub(crate) static STORAGE_IO_SIZE: Lazy<IntGaugeVec> = Lazy::new(|| {
...
+pub enum SmgrQueryType {
+    GetRelExists,
+    GetRelSize,
+    GetPageAtLsn,
+    GetDbSize,
+}
+
+#[derive(Debug)]
+pub struct SmgrQueryTimePerTimeline {
+    metrics: Vec<GlobalAndPerTimelineHistogram>, // indexed by SmgrQueryTimeOperation
```
Comment (koivunej) [inline pageserver/src/metrics.rs:613]: Vec is not needed, please look at #4892.

File: pageserver/src/metrics.rs [hunk 2]
```diff
@@ -1045,6 +1179,11 @@ impl Drop for TimelineMetrics {
             .write()
             .unwrap()
             .remove(tenant_id, timeline_id);
+
+        // The following metrics are born outside of the TimelineMetrics lifecycle but still
+        // removed at the end of it. This is because we want the smgr metrics to outlive
+        // an individual smgr connection, but not the timeline.
+
```
Comment (problame) [inline pageserver/src/metrics.rs:1186]:
```suggestion

        // The following metrics are born outside of the TimelineMetrics lifecycle but still
        // removed at the end of it. The idea is to have the metrics outlive the
        // entity during which they're observed, e.g., we the smgr metrics shall
        // outlive an individual smgr connection, but not the timeline.

```

File: pageserver/src/metrics.rs [hunk 3]
```diff
@@ -570,23 +570,160 @@ pub(crate) static STORAGE_IO_SIZE: Lazy<IntGaugeVec> = Lazy::new(|| {
...
+}
 
-pub static SMGR_QUERY_TIME: Lazy<HistogramVec> = Lazy::new(|| {
+#[derive(
+    Debug,
+    Clone,
+    Copy,
+    IntoStaticStr,
+    strum_macros::EnumCount,
+    strum_macros::EnumIter,
```
Comment (koivunej) [inline pageserver/src/metrics.rs:604]: Is this used?

Comment (problame) [reply pageserver/src/metrics.rs:604]: Yes

**Comment by problame** [comment]
Hm, is `_global` the better suffix?

File: pageserver/src/metrics.rs (hunk 3 above)
Comment (koivunej) [reply pageserver/src/metrics.rs:604]: Aah it's in the drop path and the test.

File: pageserver/src/metrics.rs [hunk 4]
```diff
@@ -570,23 +570,160 @@ pub(crate) static STORAGE_IO_SIZE: Lazy<IntGaugeVec> = Lazy::new(|| {
...
+        "Time spent on smgr query handling, aggegated by query type and tenant/timeline.",
         &["smgr_query_type", "tenant_id", "timeline_id"],
         CRITICAL_OP_BUCKETS.into(),
     )
     .expect("failed to define a metric")
 });
 
+static SMGR_QUERY_TIME_AGG: Lazy<HistogramVec> = Lazy::new(|| {
+    register_histogram_vec!(
+        "pageserver_smgr_query_seconds_agg",
```
Comment (koivunej) [inline pageserver/src/metrics.rs:632]: (Instead of replying to the _global comment).

I think this could also be `pageserver_smgr_query_total_seconds`?

**Review by koivunej — APPROVED** [review]
Name could be still adjusted but otherwise this is looking good.

File: pageserver/src/metrics.rs (hunk 4 above)
Comment (problame) [reply pageserver/src/metrics.rs:632]: I think you're misreading those naming best practices.

The _original_ per-tenant metric should have been named like so:

>  `process_cpu_seconds_total` (for an accumulating count with unit)

The `_total` is reserved for accumulating metrics like counter and, not for _agg_regating metrics like the one I'm adding here.

Going to go with `_global`, as
1. it's not part of the best practices list for naming a single metric and
2. other pre-aggregated metrics that we already have use `_global` as well.

**Pull request merged**